### PR TITLE
ETFE-3625 Private beta routing for pre validate trader

### PR DESCRIPTION
**Pull request by grahampaulcook**
Requires https://github.com/hmrc/emcs-tfe-frontend/pull/118

### CODE REVIEW

File: app/controllers/predicates/AuthActionHelper.scala [hunk 1]
```diff
@@ -35,11 +35,15 @@ trait AuthActionHelper extends BetaChecks {
...
     authorised(ern) andThen betaAllowList(betaGuard) andThen getData()
 
-  def authorisedDataRequest(ern: String)(block: DataRequest[_] => Result): Action[AnyContent] =
-    authorisedWithData(ern)(block)
+  def authorisedDataRequest(ern: String, betaGuard: Option[(String, Result)] = None)(block: DataRequest[_] => Result): Action[AnyContent] =
+    if (betaGuard.isEmpty) {
+      authorisedWithData(ern)(block)
+    } else {
+      authorisedWithBetaGuardData(ern, betaGuard.get)(block)
+    }
```
Comment (LewisMitchell) [inline app/controllers/predicates/AuthActionHelper.scala:46]: Using a fold may be better:
```suggestion
    betaGuard.fold(authorisedWithData(ern)(block))(authorisedWithBetaGuardData(ern, _)(block))
```

Comment (LewisMitchell) [reply app/controllers/predicates/AuthActionHelper.scala:46]: And may be worth applying this suggestion to the `authorisedDataRequestAsync` method

File: app/controllers/helpers/BetaChecks.scala [hunk 2]
```diff
@@ -42,4 +42,8 @@ trait BetaChecks {
 
   def changeDestinationBetaGuard(ern: String, arc: String, ver: Int)(implicit appConfig: AppConfig): Option[(String, Result)] =
     Some("tfeChangeDestination" -> Redirect(appConfig.emcsLegacyChangeDestinationUrl(ern, arc, ver)))
+
+  def preValidateBetaGuard(ern: String)(implicit appConfig: AppConfig): Option[(String, Result)] =
+    Some("tfePreValidate" -> Redirect(appConfig.emcsLegacyPreValidateUrl(ern)))
```
Comment (LewisMitchell) [inline app/controllers/helpers/BetaChecks.scala:47]: Does this have to be an option? Saves doing `.get` in the controller. I'm not sure why all of these methods are optional?

File: test/controllers/prevalidateTrader/ExciseProductCodeControllerSpec.scala [hunk 3]
```diff
@@ -57,161 +61,199 @@ class ExciseProductCodeControllerSpec extends SpecBase
...
-          val result: Future[Result] =
-            controller.onSubmit(testErn, testIndex1, NormalMode)(request.withFormUrlEncodedBody(("excise-product-code", testEpcWine)))
+      }
+      "user is NOT on the private beta list" should {
+        "redirect back to legacy" in new Fixture(userAnswers = Some(emptyUserAnswers), preValidateEnabled = false) {
+          val result = controller.onPageLoad(testErn, testIndex1, NormalMode)(request)
 
           status(result) mustEqual SEE_OTHER
-          redirectLocation(result).value mustEqual testOnwardRoute.url
+          redirectLocation(result) mustBe Some("http://localhost:8080/emcs/trader/GBWKTestErn/prevalidate")
```
Comment (LewisMitchell) [inline test/controllers/prevalidateTrader/ExciseProductCodeControllerSpec.scala:169]:
```suggestion
          redirectLocation(result).value mustEqual "http://localhost:8080/emcs/trader/GBWKTestErn/prevalidate"
```

File: test/controllers/prevalidateTrader/ExciseProductCodeControllerSpec.scala [hunk 4]
```diff
@@ -57,161 +61,199 @@ class ExciseProductCodeControllerSpec extends SpecBase
...
+      }
+      "user is NOT on the private beta list" should {
+        "redirect back to legacy" in new Fixture(userAnswers = Some(emptyUserAnswers), preValidateEnabled = false) {
+          val result: Future[Result] =
+            controller.onSubmit(testErn, testIndex1, NormalMode)(request.withFormUrlEncodedBody(("excise-product-code", testEpcWine)))
 
-      status(result) mustEqual SEE_OTHER
-      redirectLocation(result).value mustEqual routes.PrevalidateTraderStartController.onPageLoad(testErn).url
+          status(result) mustEqual SEE_OTHER
+          redirectLocation(result) mustBe Some("http://localhost:8080/emcs/trader/GBWKTestErn/prevalidate")
```
Comment (LewisMitchell) [inline test/controllers/prevalidateTrader/ExciseProductCodeControllerSpec.scala:254]:
```suggestion
          redirectLocation(result).value mustEqual "http://localhost:8080/emcs/trader/GBWKTestErn/prevalidate"
```

File: test/controllers/prevalidateTrader/PrevalidateAddToListControllerSpec.scala [hunk 5]
```diff
@@ -54,119 +58,158 @@ class PrevalidateAddToListControllerSpec extends SpecBase
...
+          val result = testController.onPageLoad(testErn)(request)
 
-        status(result) mustEqual OK
-        contentAsString(result) mustEqual view(
-          formOpt = None,
-          onSubmitCall = onSubmitCall,
-          PrevalidateAddToListHelper.addedEpcs()
-        ).toString
+          status(result) mustEqual SEE_OTHER
+          redirectLocation(result) mustBe Some("http://localhost:8080/emcs/trader/GBWKTestErn/prevalidate")
```
Comment (LewisMitchell) [inline test/controllers/prevalidateTrader/PrevalidateAddToListControllerSpec.scala:140]:
```suggestion
          redirectLocation(result).value mustEqual "http://localhost:8080/emcs/trader/GBWKTestErn/prevalidate"
```

File: test/controllers/prevalidateTrader/PrevalidateAddToListControllerSpec.scala [hunk 6]
```diff
@@ -54,119 +58,158 @@ class PrevalidateAddToListControllerSpec extends SpecBase
...
+          val result = testController.onSubmit(testErn)(req)
 
-        status(result) mustEqual BAD_REQUEST
-        contentAsString(result) mustEqual view(
-          formOpt = Some(boundForm),
-          onSubmitCall = onSubmitCall,
-          PrevalidateAddToListHelper.addedEpcs()
-        ).toString
+          status(result) mustEqual SEE_OTHER
+          redirectLocation(result) mustBe Some("http://localhost:8080/emcs/trader/GBWKTestErn/prevalidate")
```
Comment (LewisMitchell) [inline test/controllers/prevalidateTrader/PrevalidateAddToListControllerSpec.scala:208]:
```suggestion
          redirectLocation(result).value mustEqual "http://localhost:8080/emcs/trader/GBWKTestErn/prevalidate"
```

File: test/controllers/prevalidateTrader/PrevalidateTraderStartControllerSpec.scala [hunk 7]
```diff
@@ -27,56 +29,103 @@ import views.html.prevalidateTrader.PrevalidateTraderStartView
...
 
-        val request = FakeRequest(POST, routes.PrevalidateTraderStartController.onSubmit(testErn).url)
-        val result = controller.onSubmit(testErn)(request)
+          val request = FakeRequest(POST, routes.PrevalidateTraderStartController.onSubmit(testErn).url)
+          val result = controller.onSubmit(testErn)(request)
 
-        status(result) mustEqual SEE_OTHER
-        redirectLocation(result).value mustEqual routes.PrevalidateConsigneeTraderIdentificationController.onPageLoad(testErn).url
+          status(result) mustEqual SEE_OTHER
+          redirectLocation(result) mustBe Some("http://localhost:8080/emcs/trader/GBWKTestErn/prevalidate")
```
Comment (LewisMitchell) [inline test/controllers/prevalidateTrader/PrevalidateTraderStartControllerSpec.scala:127]:
```suggestion
          redirectLocation(result).value mustEqual "http://localhost:8080/emcs/trader/GBWKTestErn/prevalidate"
```

File: test/controllers/prevalidateTrader/PrevalidateTraderStartControllerSpec.scala [hunk 8]
```diff
@@ -27,56 +29,103 @@ import views.html.prevalidateTrader.PrevalidateTraderStartView
...
       }
+
+      "user is NOT on the private beta list" should {
+        "redirect back to legacy" in new Test(preValidateEnabled = false) {
+
+          val request = FakeRequest(GET, routes.PrevalidateTraderStartController.onPageLoad(testErn).url)
+          val result = controller.onPageLoad(testErn)(request)
+
+          status(result) mustEqual SEE_OTHER
+          redirectLocation(result) mustBe Some("http://localhost:8080/emcs/trader/GBWKTestErn/prevalidate")
```
Comment (LewisMitchell) [inline test/controllers/prevalidateTrader/PrevalidateTraderStartControllerSpec.scala:85]:
```suggestion
          redirectLocation(result).value mustEqual "http://localhost:8080/emcs/trader/GBWKTestErn/prevalidate"
```

File: test/controllers/prevalidateTrader/PrevalidateRemoveExciseProductCodeControllerSpec.scala [hunk 9]
```diff
@@ -48,121 +52,156 @@ class PrevalidateRemoveExciseProductCodeControllerSpec
...
-      val result = controller.onSubmit(testErn, testIndex1)(postRequest.withFormUrlEncodedBody(("value", "true")))
+    }
+    "user is NOT on the private beta list" should {
+      "redirect back to legacy" in new Fixture(Some(baseUserAnswers), preValidateEnabled = false) {
+        val result = controller.onSubmit(testErn, testIndex1)(postRequest.withFormUrlEncodedBody(("value", "true")))
 
-      status(result) mustEqual SEE_OTHER
-      redirectLocation(result).value mustEqual routes.PrevalidateTraderStartController.onPageLoad(testErn).url
+        status(result) mustEqual SEE_OTHER
+        redirectLocation(result) mustBe Some("http://localhost:8080/emcs/trader/GBWKTestErn/prevalidate")
```
Comment (LewisMitchell) [inline test/controllers/prevalidateTrader/PrevalidateRemoveExciseProductCodeControllerSpec.scala:203]:
```suggestion
          redirectLocation(result).value mustEqual "http://localhost:8080/emcs/trader/GBWKTestErn/prevalidate"
```

File: test/controllers/prevalidateTrader/PrevalidateRemoveExciseProductCodeControllerSpec.scala [hunk 10]
```diff
@@ -48,121 +52,156 @@ class PrevalidateRemoveExciseProductCodeControllerSpec
...
 
-      status(result) mustEqual SEE_OTHER
-      redirectLocation(result).value mustEqual routes.PrevalidateTraderStartController.onPageLoad(testErn).url
+    }
+    "user is NOT on the private beta list" should {
+      "redirect back to legacy" in new Fixture(Some(baseUserAnswers), preValidateEnabled = false) {
+        val result = controller.onPageLoad(testErn, testIndex1)(getRequest)
+
+        status(result) mustEqual SEE_OTHER
+        redirectLocation(result) mustBe Some("http://localhost:8080/emcs/trader/GBWKTestErn/prevalidate")
```
Comment (LewisMitchell) [inline test/controllers/prevalidateTrader/PrevalidateRemoveExciseProductCodeControllerSpec.scala:124]:
```suggestion
          redirectLocation(result).value mustEqual "http://localhost:8080/emcs/trader/GBWKTestErn/prevalidate"
```

File: test/controllers/prevalidateTrader/PrevalidateConsigneeTraderIdentificationControllerSpec.scala [hunk 11]
```diff
@@ -132,43 +181,34 @@ class PrevalidateConsigneeTraderIdentificationControllerSpec extends SpecBase wi
...
-              .returns(Future.successful(emptyUserAnswers))
-
-            val request = FakeRequest(POST, routes.PrevalidateConsigneeTraderIdentificationController.onPageLoad(testErn).url)
-            val result = controller.onSubmit(testErn)(request.withFormUrlEncodedBody("value" -> testAnswer))
-
-            status(result) mustEqual SEE_OTHER
-            redirectLocation(result).value mustEqual testOnwardRoute.url
-          }
+          status(result) mustEqual SEE_OTHER
+          redirectLocation(result) mustBe Some("http://localhost:8080/emcs/trader/GBWKTestErn/prevalidate")
```
Comment (LewisMitchell) [inline test/controllers/prevalidateTrader/PrevalidateConsigneeTraderIdentificationControllerSpec.scala:211]:
```suggestion
          redirectLocation(result).value mustEqual "http://localhost:8080/emcs/trader/GBWKTestErn/prevalidate"
```

File: test/controllers/prevalidateTrader/PrevalidateConsigneeTraderIdentificationControllerSpec.scala [hunk 12]
```diff
@@ -30,100 +32,147 @@ import views.html.prevalidateTrader.PrevalidateConsigneeTraderIdentificationView
...
           val request = FakeRequest(GET, routes.PrevalidateConsigneeTraderIdentificationController.onPageLoad(testErn).url)
           val result = controller.onPageLoad(testErn)(request)
 
-          status(result) mustEqual OK
-          contentAsString(result) mustEqual view(
-            form = formProvider(),
-            action = routes.PrevalidateConsigneeTraderIdentificationController.onSubmit(testErn)
-          )(userAnswersRequest(request), messages(request)).toString
+          status(result) mustEqual SEE_OTHER
+          redirectLocation(result) mustBe Some("http://localhost:8080/emcs/trader/GBWKTestErn/prevalidate")
```
Comment (LewisMitchell) [inline test/controllers/prevalidateTrader/PrevalidateConsigneeTraderIdentificationControllerSpec.scala:115]:
```suggestion
          redirectLocation(result).value mustEqual "http://localhost:8080/emcs/trader/GBWKTestErn/prevalidate"
```

File: app/controllers/predicates/AuthActionHelper.scala (hunk 1 above)
Comment (grahampaulcook) [reply app/controllers/predicates/AuthActionHelper.scala]: Updated at https://github.com/hmrc/emcs-tfe-frontend/pull/130/commits/97479a74f01fb8797a946d669e6dd72e6470727a

File: app/controllers/helpers/BetaChecks.scala (hunk 2 above)
Comment (grahampaulcook) [reply app/controllers/helpers/BetaChecks.scala]: It doesn't have to be an Option, only did that to save overloading the functions in AuthActionHelper although I have now made a change to that at https://github.com/hmrc/emcs-tfe-frontend/pull/130/commits/b347a231fdcd4609b83c17c82624a65e5d8e05e7

**Review by LewisMitchell — APPROVED** [review]

**Pull request merged**